### PR TITLE
workflows: generate build artifacts for CLIs

### PR DESCRIPTION
**Pull request by axbarsan**
Closes https://github.com/giantswarm/giantswarm/issues/11729

Depends on the Makefile generated with https://github.com/giantswarm/devctl/pull/63 (CLI flavour)

This PR adds a new flag to the `devctl gen workflows` command. The `-f|--flavour` flag (possible values: app|cli|library|operator)

With the `cli` flavour, in the generated github release creation workflow, there will also be a step for generating build artifacts, and adding them to the newly created release.

> 

☢️ If you want to test it, make sure you have `name` and `source` set to the correct values inside of `pkg/project.go`, in your project.

### CODE REVIEW

File: pkg/gen/input/workflows/internal/file/create_release.go [hunk 1]
```diff
@@ -189,10 +198,111 @@ jobs:
...
+      - name: Upload artifact
+        uses: actions/upload-artifact@v2
+        with:
+          name: "${{ env.BINARY }}"
+          path: "${{ env.DIR }}/${{ env.BINARY }}"
+  create_and_upload_build_artifacts:
+    name: Create and upload build artifacts
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
```
Comment (axbarsan) [inline pkg/gen/input/workflows/internal/file/create_release.go:253]: Using a matrix strategy, to build for all platforms in parallel.

File: pkg/gen/input/workflows/internal/file/create_release.go [hunk 2]
```diff
@@ -189,10 +198,111 @@ jobs:
...
       - name: Create release
+        id: create_gh_release
         uses: actions/create-release@v1
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         with:
           tag_name: "v${{ needs.gather_facts.outputs.version }}"
           release_name: "v${{ needs.gather_facts.outputs.version }}"
+
+##-ignore-left-##- if eq .CurrentFlavour .FlavourCLI##-ignore-right-##
```
Comment (kopiczko) [inline pkg/gen/input/workflows/internal/file/create_release.go:209]: When we use delimiter let's use something more sane. Maybe `{{{{` or something. The original problem was that GH actions use `${{ ... }}` and Go templates think `{{ ... }}` is it's delimiter.

File: pkg/gen/input/workflows/internal/file/create_release.go [hunk 3]
```diff
@@ -189,10 +198,111 @@ jobs:
...
+          docker cp tmp:${{ env.IMAGE_PATH }} ${{ env.DIR }}/${{ env.BINARY }}
+          docker container rm tmp
+      - name: Smoke test
+        run: |
+          ${{ env.DIR }}/${{ env.BINARY }} version
+      - name: Upload artifact
+        uses: actions/upload-artifact@v2
+        with:
+          name: "${{ env.BINARY }}"
+          path: "${{ env.DIR }}/${{ env.BINARY }}"
```
Comment (kopiczko) [inline pkg/gen/input/workflows/internal/file/create_release.go:248]: I assume this is copy/paste.

File: pkg/gen/input/workflows/internal/file/create_release.go [hunk 4]
```diff
@@ -189,10 +198,111 @@ jobs:
...
       - name: Create release
+        id: create_gh_release
         uses: actions/create-release@v1
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         with:
           tag_name: "v${{ needs.gather_facts.outputs.version }}"
           release_name: "v${{ needs.gather_facts.outputs.version }}"
+
+{{{{- if eq .CurrentFlavour .FlavourCLI}}}}
```
Comment (kopiczko) [inline pkg/gen/input/workflows/internal/file/create_release.go:209]:
```suggestion
{{{{- if eq .CurrentFlavour .FlavourCLI }}}}
```

File: pkg/gen/input/workflows/internal/file/create_release.go [hunk 5]
```diff
@@ -189,10 +198,111 @@ jobs:
...
+        env:
+          FILE_NAME: ${{ github.event.repository.name }}-${{ env.PKG_VERSION }}-${{ matrix.platform }}-amd64.tar.gz
+        with:
+          path: ${{ env.ARTIFACT_DIR }}
+          upload_url: ${{ needs.create_release.outputs.upload_url }}
+          asset_path: ${{ env.ARTIFACT_DIR }}/${{ env.FILE_NAME }}
+          asset_name: ${{ env.FILE_NAME }}
+          asset_content_type: application/octet-stream
+
+{{{{- end}}}}
```
Comment (kopiczko) [inline pkg/gen/input/workflows/internal/file/create_release.go:307]:
```suggestion
{{{{- end }}}}
```

**Pull request merged**